os/drivers/input/ist415_update.c

```diff
@@ -0,0 +1,667 @@
+		goto reset;
+	}
+
+	touchvdbg("Firmware update success\n");
+
+	return OK;
+
+reset:
+
+	ist415_reset(dev, false);
+
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ist415_check_fw
+ ****************************************************************************/
+
+int ist415_check_fw(struct ist415_dev_s *dev)
+{
+	struct firmware firmware;
+	struct ist415_fw *fw = &dev->fw;
+	int retry = 0;
+	int ret = 0;
+
+	ret = request_firmware(&firmware, FIRMWARE_PATH_WASHING_MACHINE, &dev);
+	if (ret < 0) {
+		touchdbg("Don't request firmware\n");
+		return -ENOENT;
+	}
+
+	fw->buf = (uint8_t *)kmm_malloc(firmware.size);
+	if (!fw->buf) {
+		release_firmware(&firmware);
+		touchdbg("Failed to allocate buffer\n");
+		return -ENOMEM;
+	} else {
+		memcpy(fw->buf, firmware.data, firmware.size);
+		fw->buf_size = firmware.size;
+	}
+
+	release_firmware(&firmware);
+
+	touchvdbg("Firmware loaded successfully.\n");
+
+	ret = ist415_get_update_info(dev, fw->buf, fw->buf_size);
+	if (ret < 0) {
+		kmm_free(fw->buf);
+		return ret;
+	}
+
+	fw->bin.main_ver = ist415_parse_ver(dev, FLAG_MAIN, fw->buf);
+	fw->bin.release_ver = ist415_parse_ver(dev, FLAG_RELEASE, fw->buf);
+	fw->bin.test_ver = ist415_parse_ver(dev, FLAG_TEST, fw->buf);
+	fw->bin.core_ver = ist415_parse_ver(dev, FLAG_CORE, fw->buf);
+	fw->bin.config_ver = ist415_parse_ver(dev, FLAG_CONFIG, fw->buf);
+
+	touchvdbg("Main : %08X Core: %x, Config %x, Release: %x, Test: %x\n", fw->bin.main_ver, fw->bin.core_ver, fw->bin.config_ver, fw->bin.release_ver, fw->bin.test_ver);
+
+	ret = ist415_check_auto_update(dev);
+	if (ret >= 0) {
+		kmm_free(fw->buf);
+		return OK;
+	}
+
+	touchvdbg("Update version. release(core, config, test): %x(%x, %x, %x) -> %x(%x, %x, %x)\n", fw->cur.release_ver, fw->cur.core_ver, fw->cur.config_ver, fw->cur.test_ver, fw->bin.release_ver, fw->bin.core_ver, fw->bin.config_ver, fw->bin.test_ver);
+
+	retry = IST415_RETRY_CNT;
+	while (retry--) {
+		ret = ist415_fw_update(dev, fw->buf, fw->buf_size);
+		if (ret == OK) {
+			ist415_calibrate(dev, IST415_RETRY_CNT);
+			break;
+		}
+	}
+
+	kmm_free(fw->buf);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: ist415_force_update
+ *
+ * Description
+ *   This function is called by Uart Command.
+ ****************************************************************************/
+
+int ist415_force_update(struct ist415_dev_s *dev, int argc, char **argv)
+{
+	struct firmware firmware;
+	struct ist415_fw *fw = &dev->fw;
+
+	char *path;
+	bool cal;
+	int ret;
+
+	if (argc < 4) {
+		return -EINVAL;
+	}
+
+	path = argv[2];
+	cal = atoi(argv[3]);
+
+	ret = request_firmware(&firmware, path, &dev);
+	if (ret < 0) {
+		touchdbg("do not request firmware\n");
```

Comment:
firmware->data is not released as release_firmware not called here

Reply:
Fixed!